vignettes/empirical_wls.Rmd

```diff
@@ -0,0 +1,269 @@
+
+leading to
+
+\[
+  B_{wls, i} \ne (L_i^\top)^{-1} B_{mmrm, i} L_i^\top
+\]
+
+which is contradictory with our previous result.
+Thus, these covariance estimator are identical.
+
+### Proof for Degrees of Freedom
+
+To prove
+\[
+  G_{wls, ij} = g_{wls, i}^\top \Phi g_{wls, j}
+\]
+and
+\[
+  G_{mmrm, ij} = g_{mmrm, i}^\top g_{mmrm, j}
+\]
+are identical, we only need to prove that
+
+\[
+  L^{-1} g_{wls, i} = g_{mmrm_i}
+\]
+
+where $\Phi = W^{-1}$ according to our previous expression.
+
+We first expand $L^{-1} g_{wls, i}$ and $g_{mmrm_i}$
+
+\[
+  L^{-1} g_{wls, i} = L^{-1} (I - X(X^\top W X)^{-1}X^\top W) S_i^\top A_{wls, i}^d W_i X_i (X^\top W X)^{-1} C
+\]
+
+\[
+  g_{mmrm, i} = (I - L_i^\top X(X^\top W X)^{-1}X^\top L_i) S_i^\top A_{mmrm, i}^d L_i^\top X_i (X^\top W X)^{-1} C
+\]
+
+where $S_i$ is the row selection matrix.
+
+We will prove the inner part equal
+\[
+  L^{-1} (I - X(X^\top W X)^{-1}X^\top W) S_i^\top A_{wls, i}^d W_i = (I - L^\top X(X^\top W X)^{-1}X^\top L) S_i^\top A_{mmrm, i}^d L_i^\top
+\]
+
+With the previous proof of covariance estimators, we already have
+
+\[
+  M_{wls, d, i} = W_i A_{wls, i}^d = L_i A_{mmrm, i}^d L_i^\top = M_{mmrm, d, i}
+\]
+we then need to prove
+\[
+  L^{-1} (I - X(X^\top W X)^{-1}X^\top W) S_i^\top = (I - L^\top X(X^\top W X)^{-1}X^\top L) S_i^\top L_i^{-1}
+\]
+
+and note the relationship between $(I - X(X^\top W X)^{-1}X^\top W)$ and $(I - L^\top X(X^\top W X)^{-1}X^\top L)$
+has already been proved in covariance estimator section, we only need to prove
+
+\[
+  L^{-1} (I - X(X^\top W X)^{-1}X^\top W) S_i^\top = (I - L^\top X(X^\top W X)^{-1}X^\top L) S_i^\top L_i^{-1}
+\]
+
+Apparently
+
+\[
+  L^{-1} (I - X(X^\top W X)^{-1}X^\top W) S_i^\top = L^{-1} S_i^\top - L^{-1} X(X^\top W X)^{-1}X_i^\top W_i
+\]
+
+\[
+  (I - L^\top X(X^\top W X)^{-1}X^\top L) S_i^\top L_i^{-1} = S_i^\top L_i^{-1} - L^\top X(X^\top W X)^{-1}X_i^\top
+\]
+
+And obviously
+\[
+  L^{-1} S_i^\top = S_i^\top L_i^{-1}
+\]
+
+
+\[
+  L^{-1} X(X^\top W X)^{-1}X_i W_i = L^\top X(X^\top W X)^{-1}X_i^\top
+\]
+
+because of the following
+\[
+  (X(X^\top W X)^{-1}X_i W_i)_{i} = X_i(X^\top W X)^{-1}X_i W_i \\
+  = W_i X_i(X^\top W X)^{-1}X_i^\top \\
+  = (W X(X^\top W X)^{-1}X_i^\top)_{i}
+\]
+
+# Special Considerations in Implementations
+
+## Pseudo Inverse of a Matrix
+
+Empirical covariance matrix is involved with the inverse of a matrix, or symmetric square root of a matrix.
+To calculate this, we usually requires that the matrix is positive-definite.
+However, this is not always assured in calculation.
+Thus, following @pustejovsky2018small, we use pseudo inverse to avoid this.
+We follow the following logic to obtain the pseudo inverse:
+
+1. Conduct singular value decomposition.
+2. Take square root of the reciprocals of singular values, if the value is larger than a computational threshold;
+otherwise replace the value with 0.
+3. Reconstruct the pseudo inverse matrix from modified singular values and U/V matrix.
+
```

Comment:
please also link this to the code - imagine you are a future developer and need to understand this and find the code etc.